File: src/dont_track_me/modules/certificates/auditor.py
"""TLS/SSL certificate trust store audit."""

from __future__ import annotations

import contextlib
import ssl
import time
from datetime import UTC, datetime
from typing import Any

from dont_track_me.core.base import AuditResult, Finding, ThreatLevel

# Certificate authorities that have been compromised, distrusted, or are state-controlled.
# Presence in a system trust store represents a security and privacy risk.
SUSPICIOUS_CAS: dict[str, str] = {
    "CNNIC": (
        "China Internet Network Information Center -- distrusted by Google and Mozilla "
        "after issuing unauthorized certificates via an intermediate CA."
    ),
    "WoSign": (
        "Chinese CA that issued fraudulent certificates (including for GitHub) and "
        "backdated SHA-1 certs. Distrusted by all major browsers since 2016."
    ),
    "StartCom": (
        "Israeli CA secretly acquired by WoSign. Distrusted by all major browsers "
        "after the acquisition and shared infrastructure were revealed."
    ),
    "TurkTrust": (
        "Turkish CA that accidentally issued intermediate CA certificates to "
        "organizations which were then used to create fraudulent Google certificates."
    ),
    "ANSSI": (
        "French government CA (Agence nationale de la securite des systemes d'information) "
        "that issued unauthorized certificates for Google domains in 2013."
    ),
    "Trustwave": (
        "US-based CA that issued a subordinate CA certificate to a private company "
        "for the explicit purpose of DPI/MITM traffic interception."
    ),
    "DarkMatter": (
        "UAE-based security firm linked to Project Raven surveillance operations. "
        "Applied to become a trusted root CA despite documented involvement in "
        "offensive cyber operations."
    ),
}

# Weak signature algorithms that are vulnerable to collision attacks
WEAK_ALGORITHMS: set[str] = {"md5", "sha1", "md2", "md4"}


def _parse_cert_time(time_str: str) -> float:
    """Parse a certificate time string to a Unix timestamp.

    Uses ssl.cert_time_to_seconds when available, falls back to strptime.
    """
    with contextlib.suppress(AttributeError, ValueError):
        return ssl.cert_time_to_seconds(time_str)

    # Fallback: parse manually. Format is typically "Mon DD HH:MM:SS YYYY GMT"
    for fmt in ("%b %d %H:%M:%S %Y %Z", "%b %d %H:%M:%S %Y"):
        with contextlib.suppress(ValueError):
            dt = datetime.strptime(time_str, fmt)
            return dt.replace(tzinfo=UTC).timestamp()

    # If all parsing fails, return 0 (will not be flagged as expired)
    return 0.0


def _get_org_name(cert: dict[str, Any]) -> str:
    """Extract the organization name from a certificate subject or issuer."""
    for field in cert.get("subject", ()):
        for key, value in field:
            if key == "organizationName":
                return value
    for field in cert.get("issuer", ()):
        for key, value in field:
            if key == "organizationName":
                return value
    # Fall back to commonName
    for field in cert.get("subject", ()):
        for key, value in field:
            if key == "commonName":
                return value
    return "Unknown"


def _get_common_name(cert: dict[str, Any]) -> str:
    """Extract the common name from a certificate subject."""
    for field in cert.get("subject", ()):
        for key, value in field:
            if key == "commonName":
                return value
    return "Unknown"


def _check_suspicious_ca(org_name: str, common_name: str) -> str | None:
    """Check if a certificate belongs to a known suspicious CA.

    Returns the suspicious CA key if matched, None otherwise.
    """
    combined = f"{org_name} {common_name}".lower()
    for ca_name in SUSPICIOUS_CAS:
        if ca_name.lower() in combined:
            return ca_name
    return None


def _check_weak_signature(cert: dict[str, Any]) -> str | None:
    """Check if a certificate uses a weak signature algorithm.

    The cert dict from ssl may not always include the signature algorithm directly.
    We check common fields that might indicate weak signatures.
    """
    # The Python ssl cert dict doesn't reliably expose the signature algorithm.
    # We check the serialNumber length as a heuristic: very short serial numbers
    # combined with certain issuers may indicate old/weak certs. However, the most
    # reliable check is via the 'crlDistributionPoints' or other fields.
    #
    # For the stdlib ssl module, we inspect available fields.
    # The OCSP and CRL fields sometimes hint at the algorithm, but the most
    # direct approach with stdlib is limited.
    #
    # We return None here and rely on the TLS version check and other heuristics
    # for weak algorithm detection. If future Python versions expose the
    # signatureAlgorithm field, this function can be updated.
    #
    # For now, check if any text fields mention sha1 or md5
    for field_name in ("crlDistributionPoints", "OCSP", "caIssuers"):
        field_val = cert.get(field_name, "")
        if isinstance(field_val, str):
            lower_val = field_val.lower()
            for algo in WEAK_ALGORITHMS:
                if algo in lower_val:
                    return algo
    return None


def _check_tls_versions() -> tuple[list[Finding], bool]:
    """Check which TLS versions the system supports."""
    findings: list[Finding] = []

    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)

    # Check minimum TLS version
    min_version = ctx.minimum_version

    # ssl.TLSVersion was added in Python 3.7
    tls_1_0_enabled = False
    tls_1_1_enabled = False

    with contextlib.suppress(AttributeError):
        if min_version <= ssl.TLSVersion.TLSv1:
            tls_1_0_enabled = True
        if min_version <= ssl.TLSVersion.TLSv1_1:
            tls_1_1_enabled = True

    if tls_1_0_enabled:
        findings.append(
            Finding(
                title="TLS 1.0 is enabled",
                description=(
                    "Your system allows TLS 1.0 connections, which has known vulnerabilities "
                    "including BEAST and POODLE attacks. TLS 1.0 was deprecated by RFC 8996 "
                    "in March 2021. All major browsers have disabled it."
                ),
                threat_level=ThreatLevel.HIGH,
                remediation=(
                    "Disable TLS 1.0 in your system's SSL configuration. Most applications "
                    "and browsers have already done this, but system-level libraries may "
                    "still allow it."
                ),
            )
        )

    if tls_1_1_enabled:
        findings.append(
            Finding(
                title="TLS 1.1 is enabled",
                description=(
                    "Your system allows TLS 1.1 connections. While less vulnerable than TLS 1.0, "
                    "TLS 1.1 was also deprecated by RFC 8996 and lacks modern cipher suites. "
                    "It does not support AEAD ciphers which are now considered essential."
                ),
                threat_level=ThreatLevel.HIGH,
                remediation=(
                    "Disable TLS 1.1 and ensure your system uses TLS 1.2 or TLS 1.3 as the "
                    "minimum version. TLS 1.3 is strongly preferred for its improved security "
                    "and performance."
                ),
            )
        )

    if not tls_1_0_enabled and not tls_1_1_enabled:
        findings.append(
            Finding(
                title="Modern TLS versions enforced",
                description=(
                    "Your system requires TLS 1.2 or higher, which is the current best practice. "
                    "TLS 1.3 provides additional security improvements including encrypted "
                    "handshakes and forward secrecy by default."
                ),
                threat_level=ThreatLevel.INFO,
                remediation="No action needed. Consider verifying TLS 1.3 is preferred.",
            )
        )

    return findings, tls_1_0_enabled or tls_1_1_enabled


async def audit_certificates(**kwargs: Any) -> AuditResult:
    """Audit system certificate trust stores for security and privacy issues."""
    findings: list[Finding] = []
    score = 100
    raw_data: dict[str, Any] = {}

    # --- Load system certificates ---
    certs: list[dict[str, Any]] = []
    try:
        ctx = ssl.create_default_context()
        certs = ctx.get_ca_certs(binary_form=False)
    except (ssl.SSLError, OSError):
        findings.append(
            Finding(
                title="Could not load system certificate store",
                description=(
                    "Unable to read the system's trusted CA certificate store. "
                    "This may indicate a misconfigured SSL installation."
                ),
                threat_level=ThreatLevel.MEDIUM,
                remediation="Verify your Python SSL installation and system certificate store.",
            )
        )
        return AuditResult(
            module_name="certificates",
            score=50,
            findings=findings,
            raw_data={"error": "could_not_load_certs"},
        )

    raw_data["total_ca_count"] = len(certs)
    expired_certs: list[str] = []
    suspicious_found: dict[str, list[str]] = {}
    weak_sig_certs: list[str] = []

    now = time.time()

    # --- Check 1: Trust store size ---
    if len(certs) > 200:
        findings.append(
            Finding(
                title=f"Large certificate trust store ({len(certs)} CAs)",
                description=(
                    f"Your system trusts {len(certs)} certificate authorities. A larger trust "
                    "store increases your attack surface -- each trusted CA can issue certificates "
                    "for any domain. The more CAs you trust, the more entities that could "
                    "potentially intercept your encrypted traffic via a misissued certificate."
                ),
                threat_level=ThreatLevel.MEDIUM,
                remediation=(
                    "Review your trusted CAs and remove any that you do not need. "
                    "Most users only need CAs that sign certificates for the sites they visit."
                ),
            )
        )
        score -= 5

    # --- Check 2 & 3 & 4: Per-certificate checks ---
    for cert in certs:
        org_name = _get_org_name(cert)
        common_name = _get_common_name(cert)
        cert_label = f"{org_name} ({common_name})"

        # Check for expired certificates
        not_after = cert.get("notAfter", "")
        if not_after:
            expiry_ts = _parse_cert_time(not_after)
            if expiry_ts > 0 and expiry_ts < now:
                expired_certs.append(cert_label)

        # Check for suspicious CAs
        suspicious_match = _check_suspicious_ca(org_name, common_name)
        if suspicious_match is not None:
            if suspicious_match not in suspicious_found:
                suspicious_found[suspicious_match] = []
            suspicious_found[suspicious_match].append(cert_label)

        # Check for weak signature algorithms
        weak_algo = _check_weak_signature(cert)
        if weak_algo is not None:
            weak_sig_certs.append(f"{cert_label} (uses {weak_algo})")

    # --- Report expired certificates ---
    if expired_certs:
        findings.append(
            Finding(
                title=f"Expired CA certificates found ({len(expired_certs)})",
                description=(
                    "The following trusted CA certificates have expired and should have been "
                    "removed from your trust store. Expired CAs can indicate a stale or "
                    "poorly maintained certificate store:\n"
                    + "\n".join(f"  - {c}" for c in expired_certs[:10])
                    + (
                        f"\n  ... and {len(expired_certs) - 10} more"
                        if len(expired_certs) > 10
                        else ""
                    )
                ),
                threat_level=ThreatLevel.HIGH,
                remediation=(
                    "Remove expired CA certificates from your trust store. On macOS, use "
                    "Keychain Access to find and remove them. On Linux, update your "
                    "ca-certificates package."
                ),
            )
        )
        score -= min(len(expired_certs) * 10, 40)

    # --- Report suspicious CAs ---
    for ca_name, cert_labels in suspicious_found.items():
        description = SUSPICIOUS_CAS[ca_name]
        findings.append(
            Finding(
                title=f"Suspicious CA in trust store: {ca_name}",
                description=(
                    f"{description}\n\nFound in your trust store:\n"
                    + "\n".join(f"  - {label}" for label in cert_labels)
                ),
                threat_level=ThreatLevel.CRITICAL,
                remediation=(
                    f"Remove {ca_name} certificates from your trust store. "
                    "These CAs have been distrusted by major browsers due to security "
                    "incidents or ties to surveillance operations."
                ),
            )
        )
        score -= 20

    # --- Report weak signature algorithms ---
    if weak_sig_certs:
        findings.append(
            Finding(
                title=f"Weak signature algorithms detected ({len(weak_sig_certs)} certs)",
                description=(
                    "The following CA certificates use deprecated signature algorithms "
                    "vulnerable to collision attacks:\n"
                    + "\n".join(f"  - {c}" for c in weak_sig_certs[:10])
                    + (
                        f"\n  ... and {len(weak_sig_certs) - 10} more"
                        if len(weak_sig_certs) > 10
                        else ""
                    )
                ),
                threat_level=ThreatLevel.HIGH,
                remediation=(
                    "SHA-1 and MD5 signatures are broken. An attacker with sufficient resources "
                    "can forge certificates signed with these algorithms. Remove or replace "
                    "these certificates."
                ),
            )
        )
        score -= 10

    raw_data["expired_certs"] = expired_certs
    raw_data["suspicious_cas"] = {k: v for k, v in suspicious_found.items()}
    raw_data["weak_sig_certs"] = weak_sig_certs

    # --- Check 5: TLS version support ---
    tls_findings, old_tls_enabled = _check_tls_versions()
    findings.extend(tls_findings)
    if old_tls_enabled:
        score -= 15

    # --- Check 6: Certificate Transparency (educational) ---
    findings.append(
        Finding(
            title="Certificate Transparency monitoring",
            description=(
                "Certificate Transparency (CT) is a public logging framework that records "
                "all issued TLS certificates. By monitoring CT logs, you can detect if a CA "
                "issues an unauthorized certificate for your domains. Major browsers (Chrome, "
                "Safari, Firefox) now require CT compliance for trusted certificates.\n\n"
                "You can monitor CT logs for domains you own at https://crt.sh or through "
                "Google's CT search at https://transparencyreport.google.com/https/certificates"
            ),
            threat_level=ThreatLevel.INFO,
            remediation=(
                "Enable Certificate Transparency checking in your browser. Chrome enables "
                "this by default. For domain owners, set up CT log monitoring to detect "
                "unauthorized certificates issued for your domains."
            ),
        )
    )

    score = max(0, min(100, score))

    return AuditResult(
        module_name="certificates",
        score=score,
        findings=findings,
        raw_data=raw_data,
    )

File: crates/dtm-modules/src/certificates/auditor.rs
//! TLS certificate trust store audit.

use std::collections::HashMap;

use anyhow::Result;
use dtm_core::models::{AuditOpts, AuditResult, Finding, ThreatLevel};

/// Known suspicious or distrusted certificate authorities.
const SUSPICIOUS_CAS: &[(&str, &str)] = &[
    (
        "CNNIC",
        "China Internet Network Information Center -- distrusted by Google and Mozilla \
         after issuing unauthorized certificates via an intermediate CA.",
    ),
    (
        "WoSign",
        "Chinese CA that issued fraudulent certificates (including for GitHub) and \
         backdated SHA-1 certs. Distrusted by all major browsers since 2016.",
    ),
    (
        "StartCom",
        "Israeli CA secretly acquired by WoSign. Distrusted by all major browsers \
         after the acquisition and shared infrastructure were revealed.",
    ),
    (
        "TurkTrust",
        "Turkish CA that accidentally issued intermediate CA certificates to \
         organizations which were then used to create fraudulent Google certificates.",
    ),
    (
        "ANSSI",
        "French government CA that issued unauthorized certificates for Google \
         domains in 2013.",
    ),
    (
        "DarkMatter",
        "UAE-based security firm linked to Project Raven surveillance operations. \
         Applied to become a trusted root CA despite documented involvement in \
         offensive cyber operations.",
    ),
];

/// OID for MD5 with RSA (1.2.840.113549.1.1.4).
#[cfg(feature = "certificates")]
const OID_MD5_RSA: &[u64] = &[1, 2, 840, 113549, 1, 1, 4];

/// OID for SHA-1 with RSA (1.2.840.113549.1.1.5).
#[cfg(feature = "certificates")]
const OID_SHA1_RSA: &[u64] = &[1, 2, 840, 113549, 1, 1, 5];

/// OID for MD2 with RSA (1.2.840.113549.1.1.2).
#[cfg(feature = "certificates")]
const OID_MD2_RSA: &[u64] = &[1, 2, 840, 113549, 1, 1, 2];

/// Check if a name matches any suspicious CA (case-insensitive contains).
fn check_suspicious_ca(name: &str) -> Option<(&'static str, &'static str)> {
    let lower = name.to_lowercase();
    for &(ca_name, description) in SUSPICIOUS_CAS {
        if lower.contains(&ca_name.to_lowercase()) {
            return Some((ca_name, description));
        }
    }
    None
}

/// Audit certificates with the `certificates` feature enabled.
#[cfg(feature = "certificates")]
pub async fn audit_certificates(_opts: &AuditOpts) -> Result<AuditResult> {
    use x509_parser::prelude::*;

    let mut findings = Vec::new();
    let mut score: i32 = 100;
    let mut raw_data = HashMap::new();

    // --- Load system certificates ---
    let native_certs = rustls_native_certs::load_native_certs();
    let certs = native_certs.certs;

    if certs.is_empty() {
        findings.push(Finding {
            title: "Could not load system certificate store".to_string(),
            description: "Unable to read the system's trusted CA certificate store. \
                This may indicate a misconfigured SSL installation."
                .to_string(),
            threat_level: ThreatLevel::Medium,
            remediation: "Verify your system certificate store configuration.".to_string(),
        });
        return Ok(AuditResult {
            module_name: "certificates".to_string(),
            score: 50,
            findings,
            raw_data: HashMap::from([(
                "error".to_string(),
                serde_json::json!("could_not_load_certs"),
            )]),
        });
    }

    raw_data.insert("total_ca_count".to_string(), serde_json::json!(certs.len()));

    let mut expired_certs: Vec<String> = Vec::new();
    let mut suspicious_found: HashMap<&str, Vec<String>> = HashMap::new();
    let mut weak_sig_certs: Vec<String> = Vec::new();
    let now = ::time::OffsetDateTime::now_utc();

    // --- Check 1: Trust store size ---
    if certs.len() > 200 {
        findings.push(Finding {
            title: format!("Large certificate trust store ({} CAs)", certs.len()),
            description: format!(
                "Your system trusts {} certificate authorities. A larger trust \
                 store increases your attack surface -- each trusted CA can issue \
                 certificates for any domain.",
                certs.len()
            ),
            threat_level: ThreatLevel::Medium,
            remediation: "Review your trusted CAs and remove any that you do not need.".to_string(),
        });
        score -= 5;
    }

    // --- Per-certificate checks ---
    for cert_der in &certs {
        let parsed = match X509Certificate::from_der(cert_der.as_ref()) {
            Ok((_, cert)) => cert,
            Err(_) => continue,
        };

        let issuer = parsed.issuer().to_string();
        let subject = parsed.subject().to_string();
        let cert_label = if issuer == subject {
            subject.clone()
        } else {
            format!("{subject} (issued by {issuer})")
        };

        // Check for expired certificates
        let not_after = parsed.validity().not_after.to_datetime();
        if not_after < now {
            expired_certs.push(cert_label.clone());
        }

        // Check for suspicious CAs
        let combined = format!("{subject} {issuer}");
        if let Some((ca_name, _desc)) = check_suspicious_ca(&combined) {
            suspicious_found
                .entry(ca_name)
                .or_default()
                .push(cert_label.clone());
        }

        // Check for weak signature algorithms
        let sig_oid = parsed.signature_algorithm.algorithm.iter().unwrap();
        let sig_components: Vec<u64> = sig_oid.collect();

        let weak_algo = if sig_components == OID_MD5_RSA {
            Some("MD5")
        } else if sig_components == OID_SHA1_RSA {
            Some("SHA-1")
        } else if sig_components == OID_MD2_RSA {
            Some("MD2")
        } else {
            None
        };

        if let Some(algo) = weak_algo {
            weak_sig_certs.push(format!("{cert_label} (uses {algo})"));
        }
    }

    // --- Report expired certificates ---
    if !expired_certs.is_empty() {
        let count = expired_certs.len();
        let listed: String = expired_certs
            .iter()
            .take(10)
            .map(|c| format!("  - {c}"))
            .collect::<Vec<_>>()
            .join("\n");
        let more = if count > 10 {
            format!("\n  ... and {} more", count - 10)
        } else {
            String::new()
        };

        findings.push(Finding {
            title: format!("Expired CA certificates found ({count})"),
            description: format!(
                "The following trusted CA certificates have expired and should \
                 have been removed from your trust store:\n{listed}{more}"
            ),
            threat_level: ThreatLevel::High,
            remediation: "Remove expired CA certificates from your trust store. On macOS, \
                use Keychain Access. On Linux, update your ca-certificates package."
                .to_string(),
        });
        score -= (count as i32 * 10).min(40);
    }

    // --- Report suspicious CAs ---
    for (ca_name, cert_labels) in &suspicious_found {
        let description = SUSPICIOUS_CAS
            .iter()
            .find(|(name, _)| name == ca_name)
            .map(|(_, desc)| *desc)
            .unwrap_or("Suspicious CA");

        let listed: String = cert_labels
            .iter()
            .map(|c| format!("  - {c}"))
            .collect::<Vec<_>>()
            .join("\n");

        findings.push(Finding {
            title: format!("Suspicious CA in trust store: {ca_name}"),
            description: format!("{description}\n\nFound in your trust store:\n{listed}"),
            threat_level: ThreatLevel::Critical,
            remediation: format!(
                "Remove {ca_name} certificates from your trust store. These CAs \
                 have been distrusted by major browsers due to security incidents \
                 or ties to surveillance operations."
            ),
        });
        score -= 20;
    }

    // --- Report weak signature algorithms ---
    if !weak_sig_certs.is_empty() {
        let count = weak_sig_certs.len();
        let listed: String = weak_sig_certs
            .iter()
            .take(10)
            .map(|c| format!("  - {c}"))
            .collect::<Vec<_>>()
            .join("\n");
        let more = if count > 10 {
            format!("\n  ... and {} more", count - 10)
        } else {
            String::new()
        };

        findings.push(Finding {
            title: format!("Weak signature algorithms detected ({count} certs)"),
            description: format!(
                "The following CA certificates use deprecated signature algorithms \
                 vulnerable to collision attacks:\n{listed}{more}"
            ),
            threat_level: ThreatLevel::High,
            remediation: "SHA-1 and MD5 signatures are broken. An attacker with \
                sufficient resources can forge certificates signed with these \
                algorithms. Remove or replace these certificates."
                .to_string(),
        });
        score -= 10;
    }

    raw_data.insert(
        "expired_count".to_string(),
        serde_json::json!(expired_certs.len()),
    );
    raw_data.insert(
        "suspicious_cas".to_string(),
        serde_json::json!(suspicious_found.keys().collect::<Vec<_>>()),
    );
    raw_data.insert(
        "weak_sig_count".to_string(),
        serde_json::json!(weak_sig_certs.len()),
    );

    // --- Certificate Transparency (educational) ---
    findings.push(Finding {
        title: "Certificate Transparency monitoring".to_string(),
        description: "Certificate Transparency (CT) is a public logging framework that \
            records all issued TLS certificates. By monitoring CT logs, you can detect \
            if a CA issues an unauthorized certificate for your domains. Major browsers \
            now require CT compliance for trusted certificates.\n\n\
            Monitor CT logs at https://crt.sh or Google's CT search."
            .to_string(),
        threat_level: ThreatLevel::Info,
        remediation: "Enable Certificate Transparency checking in your browser. Chrome \
            enables this by default. For domain owners, set up CT log monitoring."
            .to_string(),
    });

    let score = score.clamp(0, 100) as u32;

    Ok(AuditResult {
        module_name: "certificates".to_string(),
        score,
        findings,
        raw_data,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn check_suspicious_ca_detects_cnnic() {
        let result = check_suspicious_ca("CN=CNNIC ROOT");
        assert!(result.is_some(), "Expected CNNIC to be flagged");
        let (name, _desc) = result.unwrap();
        assert_eq!(name, "CNNIC");
    }

    #[test]
    fn check_suspicious_ca_detects_wosign() {
        let result = check_suspicious_ca("CN=WoSign CA Free SSL Certificate G2");
        assert!(result.is_some(), "Expected WoSign to be flagged");
        let (name, _desc) = result.unwrap();
        assert_eq!(name, "WoSign");
    }

    #[test]
    fn check_suspicious_ca_detects_startcom() {
        let result = check_suspicious_ca("O=StartCom Ltd.");
        assert!(result.is_some(), "Expected StartCom to be flagged");
        let (name, _desc) = result.unwrap();
        assert_eq!(name, "StartCom");
    }

    #[test]
    fn check_suspicious_ca_detects_turktrust() {
        let result = check_suspicious_ca("CN=TURKTRUST Certificate Services");
        assert!(result.is_some(), "Expected TurkTrust to be flagged");
        let (name, _desc) = result.unwrap();
        assert_eq!(name, "TurkTrust");
    }

    #[test]
    fn check_suspicious_ca_detects_anssi() {
        let result = check_suspicious_ca("O=ANSSI, CN=IGC/A");
        assert!(result.is_some(), "Expected ANSSI to be flagged");
        let (name, _desc) = result.unwrap();
        assert_eq!(name, "ANSSI");
    }

    #[test]
    fn check_suspicious_ca_detects_darkmatter() {
        let result = check_suspicious_ca("CN=DarkMatter Assured CA");
        assert!(result.is_some(), "Expected DarkMatter to be flagged");
        let (name, _desc) = result.unwrap();
        assert_eq!(name, "DarkMatter");
    }

    #[test]
    fn check_suspicious_ca_safe_names_return_none() {
        assert!(check_suspicious_ca("CN=DigiCert Global Root G2").is_none());
        assert!(check_suspicious_ca("CN=Let's Encrypt Authority X3").is_none());
        assert!(check_suspicious_ca("CN=Baltimore CyberTrust Root").is_none());
        assert!(check_suspicious_ca("CN=Amazon Root CA 1").is_none());
        assert!(check_suspicious_ca("CN=GlobalSign Root CA").is_none());
        assert!(check_suspicious_ca("").is_none());
    }

    #[test]
    fn check_suspicious_ca_case_insensitive() {
        // The function lowercases both sides, so mixed case should still match.
        assert!(check_suspicious_ca("cn=cnnic root").is_some());
        assert!(check_suspicious_ca("CN=WOSIGN CA").is_some());
        assert!(check_suspicious_ca("o=darkmatter").is_some());
    }

    #[test]
    fn test_safe_ca_no_finding() {
        // Well-known trusted CAs should NOT trigger any suspicious finding
        let safe_cas = [
            "CN=DigiCert Global Root G2",
            "CN=DigiCert High Assurance EV Root CA",
            "CN=Let's Encrypt Authority X3",
            "CN=ISRG Root X1",
            "CN=Baltimore CyberTrust Root",
            "CN=Amazon Root CA 1",
            "CN=GlobalSign Root CA",
            "CN=Comodo RSA Certification Authority",
            "CN=GeoTrust Global CA",
            "CN=VeriSign Class 3 Public Primary Certification Authority",
        ];
        for ca in safe_cas {
            assert!(
                check_suspicious_ca(ca).is_none(),
                "Expected {ca} to NOT be flagged as suspicious"
            );
        }
    }
}

/// Fallback when the `certificates` feature is not enabled.
#[cfg(not(feature = "certificates"))]
pub async fn audit_certificates(_opts: &AuditOpts) -> Result<AuditResult> {
    Ok(AuditResult {
        module_name: "certificates".to_string(),
        score: 0,
        findings: vec![Finding {
            title: "Certificate audit not available".to_string(),
            description: "The certificate audit requires the 'certificates' feature \
                to be enabled at compile time (x509-parser, rustls-native-certs)."
                .to_string(),
            threat_level: ThreatLevel::Info,
            remediation: "Rebuild with: cargo build --features certificates".to_string(),
        }],
        raw_data: HashMap::new(),
    })
}
